### DISCUSSION TITLE
v = connection.exec_driver_sql("select pg_catalog.version()").scalar() returning bytestring instead of str on Azure Postgres with psycopg>3 and sqlalchemy2

### DISCUSSION THREAD

**Discussion by ddimmich** [sqlalchemy/sqlalchemy#11373] [Usage Questions] [answered]
Hi,

We're facing an odd issue on our machines connecting to Azure Postgres - when getting the version string in 
dialects/postgresql/base.py

```
    def _get_server_version_info(self, connection):
        v = connection.exec_driver_sql("select pg_catalog.version()").scalar()
        print(v) # added by me for debugging
```

The machine connecting to the postgres server in azure prints a bytestring:

`b'PostgreSQL 15.4 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0, 64-bit'`

Whereas with the same setup on AWS's postgres (debian, same python 3.11, same installed requirement.txt) we get:

`PostgreSQL 13.13 on aarch64-unknown-linux-gnu, compiled by gcc (GCC) 7.3.1 20180712 (Red Hat 7.3.1-6), 64-bit`

Obviously the postgres versions are different, but even before upgrading to postgres 15 on Azure we still got the bytestring with Postgres 13.12 there.  The bytes type fails to be parsed by the regex, and throws the exception below.

```
  File "/data/www/venv/lib/python3.11/site-packages/sqlalchemy/dialects/postgresql/base.py", line 3198, in _get_server_version_info
    m = re.match(
        ^^^^^^^^^
  File "/usr/lib/python3.11/re/__init__.py", line 166, in match
    return _compile(pattern, flags).match(string)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: cannot use a string pattern on a bytes-like object

```

Psql otherwise works fine, and django's orm, which uses the same credentials and same postgres driver runs without issue.

When I 'hack'   this line    ` v = connection.exec_driver_sql("select pg_catalog.version()").scalar()` to have a `.decode()` at the end, sqlalchemy happily gets the version, and continues.

I've also checked if collation/encoding/ctype are any different but cannot figure out what might be causing it.  I've also tried upgrading to the latest sqlalchemy & psycopg's to no avail.  The debian machines are both running the same version and are controlled by ansible - they have the same libpq/postgres client libraries.

Not sure what else I can check.

Thank you in advance for your attention,
Damian

**Comment by CaselIT** [comment]
Hi,

This is very strange. At work I also use pg on azure and I don't have this issue.
Can you try executing `connection.exec_driver_sql("select pg_catalog.version()::text").scalar()` to see if the issue is the value returned by the db or somewhere else?

**Reply by ddimmich** [reply]
Just tried with ::text - also returns a bytestring.
```

2024-05-09 16:22:07,120 INFO sqlalchemy.engine.Engine select pg_catalog.version()::text
2024-05-09 16:22:07,120 INFO /data/www/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py _execute_context 1841 select pg_catalog.version()::text
2024-05-09 16:22:07,122 INFO sqlalchemy.engine.Engine [raw sql] {}
2024-05-09 16:22:07,122 INFO /data/www/venv/lib/python3.11/site-packages/sqlalchemy/engine/base.py _execute_context 1841 [raw sql] {}
b'PostgreSQL 15.4 on x86_64-pc-linux-gnu, compiled by gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0, 64-bit'
```

**Reply by ddimmich** [reply]
That seems to do the correct thing as well.

```
 select pg_typeof(pg_catalog.version());
 pg_typeof 
-----------
 text
(1 row)

```

**Comment by zzzeek** [comment]
hi -

sounds like a driver issue, I would test out psycopg2, pg8000 as well for comparison, then potentially report to psycopg

**Reply by ddimmich** [reply]
Thank you - will try to see if I can isolate it at that level!

**Reply by ddimmich** [reply]
https://github.com/psycopg/psycopg/issues/813 - for reference if anyone comes looking.  Was able to reproduce with psycopg directly.

**Comment by ddimmich** [comment, marked as answer]
Figured it out - on Azure the client encoding defaulted to SQL_ASCII which psycopg takes as 'no encoding' - and ends up returning byte strings.

Adding an explicit client encoding to the end of the string fixes it:

f"db+postgresql+psycopg://{DB_USERNAME}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{CELERY_RESULTS_DB}?client_encoding=utf8" 

A big thank you for taking the time to look at this @zzzeek  and @CaselIT

**Reply by zzzeek** [reply]
that's a pretty poor behavior on psycopg's part.  it should never be returning bytestrings by default under any circumstances.  if it doesnt know what to do with text it should raise

**Reply by ddimmich** [reply]
It's not ideal - luckily we got to the bottom of it after less than a day of debugging :)

https://www.psycopg.org/psycopg3/docs/api/objects.html#psycopg.ConnectionInfo.encoding - suggests that SQL_ASCII means 'no encoding' which was what led us to try setting one.  It is also odd that this would be the client encoding set by default by Azure's postgres... AWS's doesn't do that.  I may be missing something as I hadn't set that particular db server up - but I'd be surprised if our devops team had explicitly overridden these defaults.
